File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int i = 0;
        int j = height.length - 1;
        int max2 = -1;
        int max = -1;
        int water = 0;
        while(i < j){
            if (height[i] < height[j]){
                if(max < height[i]){
                    max = height[i];
                }
                else{
                    water = water + (max - height[i]);
                }
                i++;
            }
            else{
                if(max2 < height[j]){
                    max2 = height[j];
                }
                else{
                    water = water + (max2-height[j]);
                }
                j--;
            }
        }

        
        return water;
        
    }
}

File: 42-trapping-rain-water/trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # left = 0 TRIED TO DO THE O(1) APPROACH BUT IONO WHATS WRONG
        # right = len(height) - 1
        # leftmax = 0
        # rightmax = 0
        # result = 0
        # while left < right:
        #     leftmax = max(leftmax, left)
        #     rightmax = max(rightmax, right)
        #     if height[left] < height[right]:
        #         left = left + 1
        #         result = result + (leftmax - height[left])
        #     else:
        #         right = right - 1
        #         result = result + (rightmax - height[right])
        # return result
        leftmax = [0] * len(height)
        rightmax = [0] * len(height)
        currmax = 0
        result = 0
        for i in range(0, len(height)):
            leftmax[i] = currmax
            currmax = max(currmax, height[i])
        currmax = 0
        for i in range(len(height) - 1, -1, -1):
            rightmax[i] = currmax
            currmax = max(currmax, height[i])
        for i in range(0, len(height)):
            result = result + max((min(leftmax[i], rightmax[i])) - height[i], 0)
        return result

        
        
